pkg/kube/pod_runner.go

```diff
@@ -16,40 +16,118 @@ package kube
 
 import (
 	"context"
+	"io"
+	"time"
 
 	"github.com/pkg/errors"
 	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/kubernetes"
 
 	"github.com/kanisterio/kanister/pkg/consts"
 	"github.com/kanisterio/kanister/pkg/field"
 	"github.com/kanisterio/kanister/pkg/log"
 )
 
+var (
+	PodRunnerErrNotInitialized = errors.New("pod has not been initialized")
+	PodRunnerErrAlreadyStarted = errors.New("pod has already been started")
+	PodRunnerErrUnableToStart  = errors.New("unable to start pod")
+	PodRunnerErrNotStarted     = errors.New("pod not yet started")
+)
+
+// PodFileRemover provides the mechanism to remove written file from the pod.
+type PodFileRemover interface {
+	Remove(ctx context.Context) error
+	Path() string
+}
+
+type podFileRemover struct {
+	namespace     string
+	podName       string
+	containerName string
+	podWriter     PodWriter
+	path          string
+}
+
+func (pfr *podFileRemover) Remove(ctx context.Context) error {
+	return pfr.podWriter.Remove(ctx, pfr.namespace, pfr.podName, pfr.containerName)
+}
+
+func (pfr *podFileRemover) Path() string {
+	return pfr.path
+}
+
+// PodRunner allows us to start / stop pod, write file to pod and execute command within it
+type PodRunner interface {
+	PodName() string
+	Run(ctx context.Context, fn func(context.Context, *v1.Pod) (map[string]interface{}, error)) (map[string]interface{}, error)
+	StartAndWaitForPod(ctx context.Context, stopTimeout time.Duration) error
+	StopPod(ctx context.Context) error
+	WriteFileToPod(ctx context.Context, filePath string, content io.Reader) (PodFileRemover, error)
+	Exec(ctx context.Context, command []string, stdin io.Reader, stdout, stderr io.Writer) error
+}
```

Comment:
I think, this interface includes too many functions in it. We can probably split it to several smaller interfaces.

Reply:
I think we can get rid of `PodName` and return pod name from `StartAndWaitForPod` function, so that if caller is interested in pod name, it will be his concern to remember it.

And `Run` is kept for backward compatibility.